transaction/summary/src/report.rs

```diff
@@ -3,90 +3,342 @@
 pub const MAX_RECORDS: usize = MAX_OUTPUTS as usize + MAX_INPUTS as usize;
 
-/// A report of the parties and balance changes due to a transaction.
-/// This can be produced for a given TxSummary and TxSummaryUnblindingData.
+/// Maximum number of currencies with totals supported in a single report.
+///
+/// It is expected that _most_ transactions will contain one total, however,
+/// this should be large enough to support SCIs with other token types
+pub const MAX_TOTALS: usize = 4;
+
+/// A report of the parties and balance changes due to a transaction,
+/// produced for a given TxSummary and TxSummaryUnblindingData.
+///
+/// This uses a double-entry approach where outputs and totals should be
+/// balanced. For each token, totals = our inputs - sum(change outputs) ==
+/// sum(other outputs) + fee
+///
+/// SCI inputs are currently ignored
 #[derive(Clone, Debug, Default)]
-pub struct TxSummaryUnblindingReport<const RECORDS: usize = MAX_RECORDS> {
-    /// The set of balance changes that we have observed
-    // Note: We can save about 210 bytes on the stack if we store TokenId as
-    // a [u8; 8] to avoid alignment requirements. TBD if that's worth it.
-    pub balance_changes: VecMap<(TransactionEntity, TokenId), i64, RECORDS>,
-    /// The network fee that we pay
+pub struct TxSummaryUnblindingReport<
+    const RECORDS: usize = MAX_RECORDS,
+    const TOTALS: usize = MAX_TOTALS,
+> {
+    /// Transaction outputs aggregated by address and token type
+    pub outputs: Vec<(TransactionEntity, TokenId, u64), RECORDS>,
+
+    /// Total balance change for our account for each type of token in the
+    /// transaction.
+    ///
+    /// totals = our inputs - sum(change outputs)
+    ///
+    /// Note that swap inputs are elided as these are not inputs
+    /// owned by us (ie. are not spent from our account)
+    pub totals: Vec<(TokenId, TotalKind, i64), TOTALS>,
+
+    /// The network fee that we pay to execute the transaction
     pub network_fee: Amount,
+
     /// The tombstone block associated to this transaction
     pub tombstone_block: u64,
 }
 
-impl<const RECORDS: usize> TxSummaryUnblindingReport<RECORDS> {
-    /// Add value to the balance report, for some entity
-    pub fn balance_add(
-        &mut self,
-        entity: TransactionEntity,
-        token_id: TokenId,
-        value: u64,
-    ) -> Result<(), Error> {
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub enum TotalKind {
+    /// Input owned by our account
+    Ours,
+    /// Input owned by SCI counterparty
+    Sci,
+}
+
+impl<const RECORDS: usize, const TOTALS: usize> TransactionReport
+    for TxSummaryUnblindingReport<RECORDS, TOTALS>
+{
+    /// Add owned input, added to the transaction total
+    fn input_add(&mut self, amount: Amount) -> Result<(), Error> {
+        let Amount { token_id, value } = amount;
+
+        // Ensure value will not overflow
         let value = i64::try_from(value).map_err(|_| Error::NumericOverflow)?;
```

Comment:
oof, those are some large numbers... i'll have to see that we can afford to do this on the ledger, otherwise i guess i can have an 64-bit impl to run on the ledger and a 128-bit one for elsewhere...